### PR TITLE
Implement THCUNN code for GridSampler

### PR DESCRIPTION
**Pull request by zou3519**
## Summary
Implemented a THCUNN version of https://github.com/pytorch/pytorch/blob/master/torch/lib/THNN/generic/SpatialGridSamplerBilinear.c and changed GridSampler to use the new THCUNN SpatialGridSamplerBilinear when CUDNN isn't available.

This change should make it nicer for #2625 to happen.

## Test plan
  `test/run_tests.sh`

### CODE REVIEW

File: torch/nn/_functions/vision.py [hunk 1]
```diff
@@ -9,50 +9,24 @@
...
-            output = input.new(grid_sz[0], input.size(1), grid_sz[1], grid_sz[2])
-            grid = grid.contiguous()
-            if 0 in input.stride():
-                input = input.contiguous()
-            torch._C._cudnn_grid_sampler_forward(input, grid, output)
-        else:
-            backend = type2backend[type(input)]
-            output = input.new(grid_sz[0], input.size(1), grid_sz[1], grid_sz[2])
-            backend.SpatialGridSamplerBilinear_updateOutput(backend.library_state, input, grid, output)
+        backend = type2backend[type(input)]
```
Comment (soumith) [inline torch/nn/_functions/vision.py]: if cudnn is available, still use that. If it's not available, only then do you use the THCUNN implementation

**Review by soumith — CHANGES_REQUESTED** [review]
change the tests in test/test_nn.py to check both the CuDNN path and the THCUNN path. For that you can switch the `torch.backends.cudnn.enable` flag to True/False temporarily within the test context

**Comment by soumith** [comment]
additionally, can you look at some input sizes and see the performance of the cudnn version (versus) the THCUNN version.

**Comment by zou3519** [comment]
Thanks @soumith for your comments! One quick question:

by setting the flag "temporarily within the test context" do you mean flipping the flag in the test code (ie in test_grid_sample)? I tried that and I think the tests run in parallel (flipping cudnn.enable to false temporarily in test_grid_sample caused some other cudnn tests to error). 

I'm thinking about modifying the GridSampler.forward() and .backward() functions to take a "try_cudnn" parameter that defaults to cudnn.enable

**Comment by y0ast** [comment]
Awesome, very curious to see the outcome of the performance comparison with cudnn.

**Comment by zou3519** [comment]
Updated code with @soumith's comments. Performance comparison to come soon.

File: torch/nn/_functions/vision.py [hunk 2]
```diff
@@ -10,16 +10,13 @@ class GridSampler(Function):
...
-            raise RuntimeError("GridSampler needs CuDNN for processing CUDA inputs,"
-                               " but CuDNN is not enabled")
         assert cudnn.is_acceptable(input)
 
     @staticmethod
     def forward(ctx, input, grid):
         ctx.save_for_backward(input, grid)
         grid_sz = grid.size()
-        if input.is_cuda:
+        if input.is_cuda and cudnn.enabled:
```
Comment (fmassa) [inline torch/nn/_functions/vision.py]: I think it might be better to remove the `_enforce_cudnn` method, and instead check for `cudnn.is_acceptable(input)` here instead of checking for both conditions, as [`cudnn.is_acceptable` already perform those checks for you](https://github.com/pytorch/pytorch/blob/master/torch/backends/cudnn/__init__.py#L37-L56)

File: torch/lib/THCUNN/generic/THCUNN.h [hunk 3]
```diff
@@ -1012,6 +1012,18 @@ TH_API void THNN_(SpatialUpSamplingNearest_updateOutput)(
                   THCTensor *output,
                   int scale_factor);
 
+TH_API void THNN_(SpatialGridSamplerBilinear_updateOutput)(
+	               THCState *state,
```
Comment (fmassa) [inline torch/lib/THCUNN/generic/THCUNN.h]: There is a tab in here, could you fix the indentation?

File: torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+TH_API void THNN_(SpatialGridSamplerBilinear_updateGradInput)(
+    THCState *state,
+    THCTensor *input, THCTensor *gradInput,
+    THCTensor *grid, THCTensor *gradGrid,
+    THCTensor *gradOutput) {
+
+  THCUNN_assertSameGPU(state, 5, input, gradInput, grid, gradGrid, gradOutput);
+  THNN_(SpatialGridSamplerBilinear_shapeCheck)(state, input, grid, gradOutput);
+  int N = THCTensor_(size)(state, input, 0);
```
Comment (fmassa) [inline torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu]: Could you make this return a `int64_t`, to avoid implicit casting? The same for the lines below

File: torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+TH_API void THNN_(SpatialGridSamplerBilinear_updateOutput)(
+    THCState *state,
+    THCTensor *input,
+    THCTensor *grid,
+    THCTensor *output) {
+
+  THCUNN_assertSameGPU(state, 3, input, grid, output);
+  THNN_(SpatialGridSamplerBilinear_shapeCheck)(state, input, grid, NULL);
+  int N = THCTensor_(size)(state, input, 0);
```
Comment (fmassa) [inline torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu]: Same comment about the type, could you make it `int64_t` instead?

File: torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    THCState *state,
+    THCTensor *input,
+    THCTensor *grid,
+    THCTensor *gradOutput) {
+  THCUNN_argCheck(state, THCTensor_(nDimension)(state, input) == 4, 2, input,
+      "4D input tensor expected but got: %s");
+  THCUNN_argCheck(state, THCTensor_(nDimension)(state, grid) == 4, 2, grid,
+      "4D grid tensor expected but got: %s");
+
+  int nbatch   = THCTensor_(size)(state, input, 0);
```
Comment (fmassa) [inline torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu]: Same here about `int64_t`. There was a recent refactoring that replaced `long` by a fixed type.

File: torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu (hunk 4 above)
Comment (zou3519) [reply torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu]: Should I update the the int types in https://github.com/zou3519/pytorch/blob/2dfa26f1191121bcbc972385b032a3fd547ab822/torch/lib/THCUNN/SpatialGridSamplerBilinear.cu as well, or is having an int type in a __global__ function okay?

Comment (fmassa) [reply torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu]: Yeah, I think it would be better if we replaced the `int`s by `int64_t` where applicable.

**Comment by zou3519** [comment]
## Performance comparison
For a particular (N, C, IH, IW, H, W) tuple, I generated random input and grid tensors and called forward and backward.

In general, for small values of (N, C, IH, IW, H, W) (everything less than ~20), the THCUNN is
faster than the CUDNN version.

For larger sizes, the CUDNN version is roughly 2x faster.

Script: https://gist.github.com/zou3519/8e4ce6652a3dd132a8c4336e697bbbf0
Results: https://gist.github.com/zou3519/6de82f81438de1ea136658b9e9b2ced3

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
this is good stuff, looks like it's ready for merge.

**Comment by fmassa** [comment]
There are still some `int`s in the code, can we wait to see if there are compiler warnings before merge?

**Comment by soumith** [comment]
yes, sure.

**Comment by fmassa** [comment]
Is CUDA available in the build machines? I'm not seeing THC* compilation, and I saw messages like [CUDA not available, skipping tests](https://travis-ci.org/pytorch/pytorch/jobs/276885267#L3435)

**Comment by soumith** [comment]
not on travis. there's 2 more builds that'll be triggered in a few minutes, py2 and py3 builds. they have CUDA. See: https://build.pytorch.org/

**Comment by zou3519** [comment]
@fmassa I left the `int`s in [THCUNN/SpatialGridSamplerBilinear.cu](https://github.com/zou3519/pytorch/blob/2dfa26f1191121bcbc972385b032a3fd547ab822/torch/lib/THCUNN/SpatialGridSamplerBilinear.cu) because calling [THCDeviceTensor.getSize](https://github.com/pytorch/pytorch/blob/16ddc863f45290dc453b9245557908509229022f/torch/lib/THC/THCDeviceTensor.cuh#L142) returns an int.  Let me know if I should change that.

**Comment by fmassa** [comment]
@zou3519 ok, thanks!
It's interesting that `getSize` returns an `int` only, given that [`size_` can be `int64_t` as well](https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCDeviceTensor.cuh#L49-L51), but I suppose that it should be fine?
Anyway, that's how `BatchNormalization` is implemented, so I'm ok with that.
Also, one last nit: [`count` is `int64_t](https://github.com/pytorch/pytorch/pull/2737/files#diff-9c126d68263401b0b75b3405e8cb5787R58), but it's defined as [`int`](https://github.com/pytorch/pytorch/pull/2737/files#diff-81eb64f9cec28e6c00aa932646c0d88fR86). Should we care? (Sorry, I'm just being annoying here :) )

**Comment by zou3519** [comment]
@fmassa Good catch, I think I should change that. Quick question: that `int64_t` count is used as `nthreads` in [CUDA_KERNEL_LOOP](https://github.com/zou3519/pytorch/blob/2dfa26f1191121bcbc972385b032a3fd547ab822/torch/lib/THCUNN/SpatialGridSamplerBilinear.cu#L98). Unfortunately that macro [uses it as an int](https://github.com/pytorch/pytorch/blob/0f65c9267d5ec55584b0ec65acb5374c95af9c16/torch/lib/THCUNN/common.h). 

Should I just change [`count`](https://github.com/zou3519/pytorch/blob/aea66d237087839026bfc4e84a08c6577551879f/torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu#L87) back to an `int` for consistency?

**Comment by fmassa** [comment]
@zou3519 might be better to change it to `int` for now (by a cast for example), but that might be something we might look again in the future. Not sure if we want to be using `int`s there

File: torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  int64_t W = THCTensor_(size)(state, grid, 2);
+
+  // resize output to the same shape as input
+  THCTensor_(resize4d)(state, output, N, C, H, W);
+
+  THCDeviceTensor<real, 4> devInput = toDeviceTensor<real, 4>(state, input);
+  THCDeviceTensor<real, 4> devGrid = toDeviceTensor<real, 4>(state, grid);
+  THCDeviceTensor<real, 4> devOutput = toDeviceTensor<real, 4>(state, output);
+
+  int64_t count = N*H*W*2;
```
Comment (fmassa) [inline torch/lib/THCUNN/generic/SpatialGridSamplerBilinear.cu]: nit: this is also an `int64_t`.

**Comment by soumith** [comment]
this is lovely! thanks @zou3519 !
